### PR TITLE
[Ready] Order in select and select2

### PR DESCRIPTION
**Pull request by eduardoarandah**
Select and Select2 controls don't have order, which is annoying. 
This commit fixes it

## before
imagen

## after
imagen

### CODE REVIEW

**Comment by eduardoarandah** [comment]
I’m trying to imagine a case when somebody needs to order by a field they don’t see. 

Never in my life have seen a dropdown sorted by other than alphabetical

**Comment by eduardoarandah** [comment]
Is it possible to add a “method” to be called in the related model?
Relations can be restricted ordered as much as we want !

**Comment by AbbyJanke** [comment]
Examples:

- Logs, you display the log message but want it sorted by ID so you know which came first or last.
- “Account Credits”, you apply financial credits to a customer say you have several credits of the same amount. Need to know which one is which for sure.

**Comment by eduardoarandah** [comment]
log messages in a dropdown list, unsorted?

**Comment by AbbyJanke** [comment]
Not unsorted, sorted by the “created_at” date rather then the “message” which is shown to the user. Its a small niche case i admit but its one i use within a current application i have.

**Comment by AbbyJanke** [comment]
I’m not saying we shouldnt make them sortable, i’m simply saying we have an option to customize how they are sorted. So expanding on my first comment, default sorr by the attribute being displayed but allow an override.

**Comment by OliverZiegler** [comment]
@eduardoarandah Thanks for your effort but there should not be a default in the field I think.

We for example use global scope on the model itself for sorting or add Subclasses to models with global scope that orders them for our selects.
With your approach you would overwrite this functionality and I suppose many larval developers define global scopes for default order of models.

So we should try to provide custom sorting via additional option (so this will also be non-breaking).

For custom Scopes this references #1266 and brings also possibility to better controls which models can be selected.

For sorting there is #1112 which suggests using an approach like the (relatively new) searchLogic attribute.

**Comment by eduardoarandah** [comment]
What do you think doing it in a 80/20 approach?

**Case a)**
80% of the time select and select2 are used for small catalogs, like "category", "city", "user_type", "post_type" that need sorting by its display attribute.

**Case b)**
20% of the time, data must be scoped and sorted in custom ways.

**Docs could be like this:**

**select**
```
[  // Select
   'label' => "Category",
   'type' => 'select',
   'name' => 'category_id', // the db column for the foreign key
   'entity' => 'category', // the method that defines the relationship in your Model
   'attribute' => 'name', // foreign key attribute that is shown to user
   'model' => "App\Models\Tag" // foreign key model
]
Notes: 
if you don't want it sorted alphabetically add
 'sorted' => false 

if you want custom data see field "select_from_array"
```

**select_from_array**

```
[ // select_from_array
    'name' => 'template',
    'label' => "Template",
    'type' => 'select_from_array',
    'options' => [‘one’ => ‘One’, ‘two’ => ‘Two’],
    'allows_null' => false,
    'default' => 'one',
    // 'allows_multiple' => true, // OPTIONAL; needs you to cast this to array in your model;
],
Note: if you want data from a model you can do something like this:
'options'=> \App\Models\Category::whereActive(1)->orderBy('name')->pluck('name', 'id')

```

What do you think?

@OliverZiegler I didn't know about scopes and search logic, they sound great but I prefer to keep it simple. 
A scope can also be used in **select_from_array** like this:

In your model
```
public function scopeActive($query)
    {
        return $query->where('active', 1)->orderBy('name');
    }
```

in a select_from_array field:

```
[ 
    'name' => 'template',
    'label' => "Template",
    'type' => 'select_from_array',
     'options'=> \App\Models\Category::active()->pluck('name', 'id')
],
```

**Comment by OliverZiegler** [comment]
@eduardoarandah well I don't agree with your case *a)*.

It's definitely bad to force ordering by display attribute and we should not apply more and more booleans fore "easy" extension but come up with a good and widely usable solution ;)

Beside this, your suggested solution is kind of a breaking-change, as the current order will be different for some users if they don't update their field description and add this `'sort' => false` entry.

Furthermore your solution with select_from_array will not work like this as you need a `key=>value` array I think.

**Comment by eduardoarandah** [comment]
@AbbyJanke @OliverZiegler 

Ok, check this out: 

**Sorted by default, unless 'sorted' => false**

https://github.com/eduardoarandah/CRUD/commit/a4eb797753492f7c27958badbb2ddabb5fe30c5f

For more complex cases, just use select_from_array with example provided

**Comment by AbbyJanke** [comment]
I would suggest the opposite if `sorted` is true then sort it. Otherwise if somebody expects their current selects to be a certain way and then we make this change it would mess up their application. I wouldn't call it a breaking change, but change enough to confuse folks.

```php
if(isset($field['sorted']) && $field['sorted'] == true)
    $models=$field['model']::orderBy($field['attribute'])->get();
else
    $models=$field['model']::all();
```

**Comment by eduardoarandah** [comment]
Ok, this one wouldn't break anybody's code:
(sorted=false by default)

https://github.com/eduardoarandah/CRUD/commit/5f1496d292f227d5c6b0b2fe2eeae19d430a6b05

And for the special cases I think this example could help in the docs:

```
**select_from_array** 
[ 
    'name' => 'template',
    'label' => "Template",
    'type' => 'select_from_array',
     'options'=> \App\Models\Category::active()->pluck('name', 'id')
],
```

**Comment by OliverZiegler** [comment]
Still thinking it's not good to introduce boolean flags but instead provide good implementation for this to be more widely usable...
As soon as on makes changes to provide custom sorting here, this will become a huge mess in the blade oder will be breaking as we remove this again... 🤔 

@tabacitu please tell us your opinion

**Comment by tswonke** [comment]
I totally agree with @OliverZiegler

File: src/CrudPanel.php [hunk 1]
```diff
@@ -66,6 +67,10 @@ class CrudPanel
     // TONE FIELDS - TODO: find out what he did with them, replicate or delete
     public $sort = [];
 
+    public $custom_form_view='';
```
Comment (lloy0076) [inline src/CrudPanel.php]: WHY must there be no PHP Doc?

File: src/CrudPanel.php [hunk 2]
```diff
@@ -66,6 +67,10 @@ class CrudPanel
...
+    public $wide_form=false;
```
Comment (lloy0076) [inline src/CrudPanel.php]: No PHP Doc?

File: src/PanelTraits/CustomFormView.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
+<?php
+
+namespace Backpack\CRUD\PanelTraits;
+
+trait CustomFormView
+{
+    //when set as wide, removes the from wrapper that makes it look narrow
+    public function setFormView($custom_form_view, $is_wide = false)
```
Comment (lloy0076) [inline src/PanelTraits/CustomFormView.php]: No PHP Doc?

File: src/PanelTraits/CustomFormView.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    {
+        $this->custom_form_view = $custom_form_view;
+        $this->wide_form        = $is_wide;
+    }
+
+    public function extractFields($fields)
```
Comment (lloy0076) [inline src/PanelTraits/CustomFormView.php]: @#@##$#$ Still no PHP Doc?

File: src/PanelTraits/CustomFormView.php [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    {
+        $f = [];
```
Comment (lloy0076) [inline src/PanelTraits/CustomFormView.php]: `$f` is not a good name.

File: src/PanelTraits/CustomFormView.php [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        foreach ($fields as $field) {
+            $f[$field['name']] = $field;
+        }
+        return $f;
+    }
+    public function extractTemplates($fields)
```
Comment (lloy0076) [inline src/PanelTraits/CustomFormView.php]: PHP Doc?

File: src/PanelTraits/CustomFormView.php [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    {
+        $t = [];
```
Comment (lloy0076) [inline src/PanelTraits/CustomFormView.php]: `$t` is not a good name.

File: src/resources/views/create.blade.php [hunk 8]
```diff
@@ -15,8 +15,12 @@
 @endsection
 
 @section('content')
+
+{{-- this is the form wrapper, we can set it off to let the programmer use the hole screen --}}
```
Comment (lloy0076) [inline src/resources/views/create.blade.php]: Donuts have holes. But one eats whole donuts. Except the holes...

File: src/resources/views/edit.blade.php [hunk 9]
```diff
@@ -15,8 +15,12 @@
 @endsection
 
 @section('content')
+
+{{-- this is the form wrapper, we can set it off to let the programmer use the hole screen --}}
```
Comment (lloy0076) [inline src/resources/views/edit.blade.php]: s/ hole / whole /

File: src/resources/views/fields/select.blade.php [hunk 10]
```diff
@@ -16,7 +16,13 @@
         @endif
 
         @if (isset($field['model']))
-            @foreach ($field['model']::all() as $connected_entity_entry)
+            @php        
+                if(isset($field['sorted']) && $field['sorted'] == true)
```
Comment (lloy0076) [inline src/resources/views/fields/select.blade.php]: `===`?

File: src/resources/views/fields/select2.blade.php [hunk 11]
```diff
@@ -18,7 +18,13 @@
         @endif
 
         @if (isset($field['model']))
-            @foreach ($field['model']::all() as $connected_entity_entry)
+            @php        
+                if(isset($field['sorted']) && $field['sorted'] == true)
```
Comment (lloy0076) [inline src/resources/views/fields/select2.blade.php]: `===`?

File: src/resources/views/form_content.blade.php [hunk 12]
```diff
@@ -2,8 +2,13 @@
 <input type="hidden" name="locale" value={{ $crud->request->input('locale')?$crud->request->input('locale'):App::getLocale() }}>
 @endif
 
-{{-- See if we're using tabs --}}
-@if ($crud->tabsEnabled())
+@if($crud->custom_form_view)
+{{-- if we are using a custom form view extract template name and field for the programmer --}}
+    @include($crud->custom_form_view,[
+      't'=>$crud->extractTemplates($fields),
```
Comment (lloy0076) [inline src/resources/views/form_content.blade.php]: Ok, humour me - what is `t` and `f` here?

File: src/resources/views/form_content.blade.php [hunk 13]
```diff
@@ -2,8 +2,13 @@
...
+      'f'=>$crud->extractFields($fields)
+      ])
+@elseif ($crud->tabsEnabled()){{-- See if we're using tabs --}}
```
Comment (lloy0076) [inline src/resources/views/form_content.blade.php]: Is that comment a useless comment?

File: src/resources/views/form_content.blade.php (hunk 12 above)
Comment (AbbyJanke) [reply src/resources/views/form_content.blade.php]: based on looking at the file, its an array where `t` is short fort `templates` and `f` is short for `fields`.

**Comment by tabacitu** [comment]
Wow... Such a great conversation here... Sorry to have missed this, guys... must have slipped through the cracks...

I agree with the end conclusion here (if I understand it correctly):
- we _should_ provide a way to order/sort the ```select```, ```select2``` fields;
- we _should NOT_ apply that order by default, because we'd break people's apps and/or prevent them from using global scopes efficiently;

 What I do NOT agree with is the implementation. I think rather than ```sort``` we can give the developer the ability to change the ```options``` or ```query``` (however you want to put it). This would not only allow them to apply a certain order, but also apply a local scope, or filter the entries. Check out the PR for it here please https://github.com/Laravel-Backpack/CRUD/pull/1694 . I think it's a great way to solve multiple problems.

I do agree with @eduardoarandah that this blurrs the lines between ```select``` and ```select_from_array```, though, but... I still think it's the best way to proceed. They're a little different - one uses arrays, one uses Eloquent objects. So I think each of them still has its place.

Thanks a lot for this back-and-forth, looks like together you got to a great conclusion, and it was very useful. I just _think_ we now have a better implementation of that conclusion (again, here https://github.com/Laravel-Backpack/CRUD/pull/1694). Let me know if you think I'm wrong and we'll reopen this implementation.

Thanks again. Cheers!

**Pull request closed without merging**